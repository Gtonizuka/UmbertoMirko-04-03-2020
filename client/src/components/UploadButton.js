import React, { useState, useContext } from 'react';
import './UploadButton.scss';
import axios from 'axios';
import { FilesContext } from '../context/FilesContext';
import { AlertContext } from '../context/AlertContext';

const UploadButton = () => {
  const { addFile } = useContext(FilesContext);
  const { updateAlert } = useContext(AlertContext);

  const [uploadedFile, setUploadedFile] = useState({});
  const [fileName, setFileName] = useState('');
  const [active, setActive] = useState(false);

  function handleFile(e) {
    const file = e.target.files[0];
    // Avoid running function if a file is selected and then user select another file and cancel it
    if (file) {
      const mbSize = e.target.files[0].size / 1024 / 1024; // in MB
      if (mbSize > 10) {
        updateAlert('File size exceeds 10 MB', 'WARNING');
        return;
      } else {
        setUploadedFile(file);
        setFileName(file.name);
        setActive(true);
      }
    }
  }

  // Upload file handler
  const fileUploadHandler = async () => {
    const formData = new FormData();
    formData.append('imageFe', uploadedFile, fileName);
    try {
      const res = await axios.post('/api/upload', formData, {
        headers: {
          'content-type': 'multipart/form-data'
        }
      });
      if (res.status === 200) {
        updateAlert('File has been uploaded succesfully', 'SUCCESS');

        const { _id, name, size } = res.data;
        addFile(_id, name, size);
        setActive(false);
        return;
      }
      updateAlert(
        'Problem with uploading your file. Please try again.',
        'WARNING'
      );
      setActive(false);
    } catch (err) {
      updateAlert(
        `Problem with uploading your file. ${err.response.data}`,
        'WARNING'
      );
      // throw new Error(err);
    }
  };

  return (
    <>
      <input type='file' onChange={handleFile} accept='image/jpeg, image/png' />
      <button onClick={fileUploadHandler} disabled={!active}>
        UPLOAD
      </button>
    </>
  );
};

export default UploadButton;
