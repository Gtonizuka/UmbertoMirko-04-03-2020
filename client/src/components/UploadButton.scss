button {
  text-transform: uppercase;
}

.custom-file-input {
  color: transparent;
  display: flex;
  width: 140px;
}
.custom-file-input::-webkit-file-upload-button {
  visibility: hidden;
}
.custom-file-input::before {
  content: 'Select some files';
  color: black;
  display: inline-block;
  background: -webkit-linear-gradient(top, #f9f9f9, #e3e3e3);
  border: 1px solid #999;
  border-radius: 3px;
  padding: 5px 8px;
  outline: none;
  white-space: nowrap;
  -webkit-user-select: none;
  cursor: pointer;
  text-shadow: 1px 1px #fff;
  font-weight: 700;
  font-size: 10pt;
}
.custom-file-input:hover::before {
  border-color: black;
}
.custom-file-input:active {
  outline: 0;
}
.custom-file-input:active::before {
  background: -webkit-linear-gradient(top, #e3e3e3, #f9f9f9);
}

.upload-meta {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  margin-top: 10px;
  max-width: 250px;
  width: 250px;

  button {
    background-color: #456bd9;
    border-color: #3e60c3;
    border-radius: 6px;
    box-sizing: border-box;
    color: #fff;
    float: left;
    font-size: 1em;
    font-weight: 600;
    padding: 0.5em 1em;
    width: 130px;
  }

  button:not(disabled) {
    cursor: pointer;
    opacity: 1;
    transition-duration: 0.3s;
    transition-property: background-color, border-color, color;
    transition-timing-function: ease;
  }
  button:disabled {
    background-color: #456bd9;
    opacity: 0.4;
    cursor: not-allowed;
  }

  .file-name {
    flex: 1 1 100%;
    font-size: 0.75em;
    margin-top: 10px;
    text-align: left;
  }

  // MOBILE
  @media only screen and (max-width: 759px) {
    max-width: 100%;
    width: 100%;
    button {
      width: 100%;
    }
  }
}
