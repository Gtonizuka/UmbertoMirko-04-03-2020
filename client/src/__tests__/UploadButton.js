import React from 'react';
import { mount } from 'enzyme';
import UploadButton from '../components/UploadButton';
import FilesContextProvider, { FilesContext } from '../context/FilesContext';
import AlertContextProvider from '../context/AlertContext';
import { findByTestAttr } from '../utils/testUtils';

describe('Upload', () => {
  let wrapper;
  beforeEach(() => {
    const contextValues = { text: 'mock', msg: 'SUCCESS' };

    // Below mounting is  needed as Enzyme does not yet support shallow mocks
    wrapper = mount(
      <FilesContextProvider>
        <AlertContextProvider value={contextValues}>
          <UploadButton />
        </AlertContextProvider>
      </FilesContextProvider>
    );
  });

  test('should be defined', () => {
    expect(wrapper).toBeDefined();
  });

  test('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should render a file input', () => {
    const element = findByTestAttr(wrapper, 'file-input');
    expect(element.length).toBe(1);
  });

  it('should have an input value', () => {
    expect(typeof wrapper.find('.custom-file-input').instance().value).toBe(
      'string'
    );
  });

  it('should render button value', () => {
    const element = findByTestAttr(wrapper, 'button');
    expect(element.length).toBe(1);
    expect(element.text()).toEqual('UPLOAD');
  });
});
