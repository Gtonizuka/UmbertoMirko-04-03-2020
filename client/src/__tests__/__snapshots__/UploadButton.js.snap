// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload Button should be defined 1`] = `ReactWrapper {}`;

exports[`Upload Button should match snapshot 1`] = `ReactWrapper {}`;

exports[`Upload should match snapshot 1`] = `ReactWrapper {}`;
